cryodrgn/source.py

```diff
@@ -0,0 +1,278 @@
+        # These are traditionally available as the 'fname' attribute of the LazyImage class, hence only used by
+        # calling code when lazy=True
+        if filenames is None:
+            filenames = [""] * self.n
+        elif isinstance(filenames, str):
+            filenames = [filenames] * self.n
+        else:
+            assert len(filenames) == self.n
+            filenames = list(filenames)
+        self.filenames = np.array(filenames)
+
+        self.n_workers = n_workers
+
+    def __len__(self):
+        return self.n
+
+    def __getitem__(self, indices: Union[int, np.ndarray, slice]) -> np.ndarray:
+        if np.isscalar(indices):
+            indices = np.array([indices])
+        elif isinstance(indices, Iterable):
+            indices = np.fromiter(indices, int)
+        elif isinstance(indices, slice):
+            start, stop, step = indices.start, indices.stop, indices.step
+            start = start or 0
+            stop = min(stop, self.n) if stop is not None else self.n
+            step = step or 1
+            assert (
+                start >= 0 and stop >= 0 and step >= 0
+            ), "Only positive start/stop/step supported"
+            indices = np.arange(start, stop, step)
+        else:
+            raise TypeError("Unsupported Type for indices")
+
+        if np.any(indices >= self.n):
+            raise ValueError(f"indices should be < {self.n}")
+
+        images = self._images(indices)
+        if images.ndim == 3 and images.shape[0] == 1:
+            images = images.squeeze(axis=0)
+
+        return images
+
+    def _images(self, indices: np.ndarray):
+        raise NotImplementedError("Subclasses must implement this")
+
+
+class MRCFileSource(ImageSource):
+    def __init__(self, filepath: str, *args, **kwargs):
+        header = MRCHeader.parse(filepath)
+        self.header = header
+        self.mrcfile_path = filepath
+        self.dtype = header.dtype
+        self.start = 1024 + header.fields["next"]  # start of image data
+        self.nz, self.ny, self.nx = (
+            header.fields["nz"],
+            header.fields["ny"],
+            header.fields["nx"],
+        )
+        assert self.ny == self.nx, "Only square images supported"
+        self.shape = self.nz, self.ny, self.nx
+        self.stride = self.dtype().itemsize * self.ny * self.nx
+
+        super().__init__(L=self.ny, n=self.nz, filenames=filepath)
+
+    def _images(self, indices: np.ndarray) -> np.ndarray:
+        # TODO: Load array in contiguous chunks instead of using an iterative loop
+        with open(self.mrcfile_path) as f:
+            data = np.empty((len(indices), self.ny, self.nx))
+            for i, index in enumerate(indices):
+                # TODO: Do this concurrently in several threads
+                f.seek(
+                    self.start
+                )  # TODO: Do not seek in every iteration - find a better way!
```

Comment:
This looks like it's only doing lazy loading, I assume you're still working on eager loading?

Reply:
This specific part of the code is doing eager-loading because it has to (the caller is asking for data). The `source.py` code is complete in terms of lazy/eager loading (though the interface may change a per your other comments in this thread).